### ISSUE TITLE
Store confidences in neural networks output

### ISSUE THREAD

**Issue by lisitsyn** [shogun-toolbox/shogun#2758]
There is no notion of confidences in neural networks at the moment which is baad.

**Comment by karlnapf** [comment]
Maybe that is not so bad. NN's notion of confidence is neither calibrated nor stable. Using them to quantify uncertainty usually is dangerous. Or has this changed?

**Comment by lisitsyn** [comment]
Well softmax is pretty calibrated.

I met this issue when applied to some classification when ranking is important. SVM outputs some data which is not calibrated as well but you can extract ranking

**Comment by karalets** [comment]
Guys,

Feedforward NN do not understand uncertainty on their parameters. The only thing they can do is quantify uncertainty on final output variables, but is is a maximum likelihood or MAP estimate which has previously thrown away all the uncertainty between inputs and that.
As such, the output (i.e. of a softmax or the probability of binary unit for classification or the variance of a gaussian) is the error bar of the NN in terms of its outputs. This does not need to be read out explicitly in any further way, really, it is part of the output parameters.

**Comment by khalednasr** [comment]
The outputs of the network can very well be interpreted as confidences in the case of multiclass classification using softmax or logistic units. I guess we could add that, since implementing it would be very easy.
